app/ethWallet-geth.js

```diff
@@ -0,0 +1,258 @@
+const fs = require('fs-extra')
+const path = require('path')
+const dns = require('dns-then')
+const {spawn} = require('child_process')
+const portfinder = require('portfinder')
+const net = require('net')
+
+const {app, ipcMain} = require('electron')
+const {getExtensionsPath} = require('../js/lib/appUrlUtil')
+const appStore = require('../js/stores/appStore')
+const ledgerState = require('./common/state/ledgerState')
+
+const envNet = process.env.ETHEREUM_NETWORK
+const gethDataDir = path.join(app.getPath('userData'), envNet || 'ethereum')
+
+const gethProcessKey = process.platform === 'win32'
+  ? 'geth.exe'
+  : 'geth'
+const ipcPath = process.platform === 'win32'
+  ? '\\\\.\\pipe\\geth.ipc'
+  : path.join(gethDataDir, 'geth.ipc')
+const pidPath = process.platform === 'win32'
+  ? '\\\\.\\pipe\\geth.pid'
+  : path.join(gethDataDir, 'geth.pid')
+const gethProcessPath = path.join(getExtensionsPath('bin'), gethProcessKey)
+
+const configurePeers = async (dataDir) => {
+  const discoveryDomain = `_enode._tcp.${envNet || 'mainnet'}.ethwallet.bravesoftware.com`
+  const newNodes = await dns.resolveSrv(discoveryDomain)
+  const newNodesNames = newNodes.map(({ name }) => name)
+
+  // start without await to take advantage of async parallelism
+  const newNodesPublicKeysPromises = Promise.all(newNodesNames.map(name => dns.resolveTxt(name)))
+  const newNodesIps = await Promise.all(newNodesNames.map(name => dns.resolve4(name)))
+  const newNodesPublicKeys = await newNodesPublicKeysPromises
+
+  const enodes = newNodes.map(({name, port}, i) => `enode://${newNodesPublicKeys[i]}@${newNodesIps[i]}:${port}`)
+
+  await fs.writeFile(path.join(dataDir, 'static-nodes.json'), JSON.stringify(enodes))
+}
+
+// needs to be shared to the eth-wallet app over ipc
+let wsPort
+// needs to be shared to the metamask extension
+let rpcPort
+
+let geth
+let gethProcessId
+let gethRetryTimeoutId
+const gethRetryInterval = 30000
+
+const spawnGeth = async () => {
+  portfinder.basePort = 40400
+  const port = await portfinder.getPortPromise()
+
+  portfinder.basePort = 40600
+  wsPort = await portfinder.getPortPromise()
+
+  portfinder.basePort = 40800
+  rpcPort = await portfinder.getPortPromise()
+
```

Comment:
do either the RPC or websocket ports accept connections from `localhost` or other whitelisted origins (not counting `chrome-extension://...`)? if so we need to make sure the RPC and websocket ports are not vulnerable to DNS rebinding: https://github.com/transmission/transmission/pull/468

Reply:
the solution is to have the listening service validate the incoming request's `host` header before doing any processing on it. here is an example for webtorrent: https://github.com/webtorrent/webtorrent/pull/1260/files

Reply:
would setting the following options be sufficient?

```
  --rpccorsdomain value  Comma separated list of domains from which to accept cross origin requests (browser enforced)
  --rpcvhosts value      Comma separated list of virtual hostnames from which to accept requests (server enforced). Accepts '*' wildcard. (default: "localhost")
```

We should set these after we get more concrete metamask integration?

Reply:
@Slava CORS is unrelated. the latter looks like it should do the right thing but i haven't audited the code.

Reply:
`--rpcvhosts` does the right thing for HTTP, but I don't think it does the right thing for WS. looking into it more...

https://github.com/ethereum/go-ethereum/blob/5d7e18539e32cb4f1aafab8e977e28a7cd34da9c/rpc/http.go#L314